src/openforms/registrations/contrib/json_dump/plugin.py

```diff
@@ -95,34 +96,107 @@ def process_variables(submission: Submission, values: JSONObject):
                 # not relevant here
                 continue
 
-            component = get_component(variable)
-            if component is None or component["type"] != "file":
-                continue
-
-            encoded_attachments: list[JSONValue] = [
-                {
-                    "file_name": attachment.original_name,
-                    "content": encode_attachment(attachment),
-                }
-                for attachment in submission.attachments
-                if attachment.form_key == key
-            ]
-
-            match (
-                component.get("multiple", False),
-                n_attachments := len(encoded_attachments),
-            ):
-                case False, 0:
-                    values[key] = None
-                case False, 1:
-                    values[key] = encoded_attachments[0]
-                case True, int():
-                    values[key] = encoded_attachments
-                case False, int():  # pragma: no cover
-                    raise ValueError(
-                        f"Cannot have multiple attachments ({n_attachments}) for a "
-                        "single file component."
-                    )
+            component = get_component(variable, submission)
+            assert component is not None
+
+            match component["type"]:
+                case "file":
+                    encoded_attachments = [
+                        {
+                            "file_name": attachment.original_name,
+                            "content": encode_attachment(attachment),
+                        }
+                        for attachment in submission.attachments
+                        if attachment.form_key == key
+                    ]
+
+                    base_schema = {
+                        "title": component["label"],
+                        "type": "object",
+                        "properties": {
+                            "file_name": {"type": "string"},
+                            "content": {"type": "string", "format": "base64"},
+                        },
+                        "required": (
+                            ["file_name", "content"]
+                            if len(encoded_attachments) != 0
+                            else []
+                            # No required properties when there are no attachments
+                        ),
+                        "additionalProperties": False,
+                    }
+
+                    match (
+                        component.get("multiple", False),
+                        n_attachments := len(encoded_attachments),
+                    ):
+                        case False, 0:
```

Comment:
So you can do much more advanced things with pattern matching, e.g.:

```py
match component:
    case {"type": "file", multiple: False}:
        ...
    case {"type": "file", multiple: True}:
        ...
    case {"openForms": {"dataSrc": "variable"}, "type": "radio" | "selectboxes"}:
        ...
    case {"openForms": {"dataSrc": "variable"}, "type": "select"}:
        ...
```
etc.

See https://github.com/open-formulieren/open-forms/blob/19a3c8af294d26da4c8f2be1f49d06d9083e23e5/bin/report_component_problems.py#L21 for a real example.

It can make the code a lot flatter/easier to follow.

Reply:
You raise some good points 🤔 you probably can add another dict without multiple key to the multiple False match since they're semantically the same, using the pipe symbol